src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java

```
@@ -0,0 +1,153 @@
+     * <b>{@code null} may not be returned, or an exception will be thrown!</b>
+     */
+    @ApiStatus.OverrideOnly
+    protected abstract T createSnapshot();
+
+    /**
+     * Roll back to a state previously created by {@link #createSnapshot}.
+     */
+    @ApiStatus.OverrideOnly
+    protected abstract void revertToSnapshot(T snapshot);
+
+    /**
+     * Signals that the snapshot will not be used anymore, and is safe to cache for future calls to {@link #createSnapshot},
+     * or discard entirely.
+     */
+    @ApiStatus.OverrideOnly
+    protected void releaseSnapshot(T snapshot) {}
+
+    /**
+     * Update the stored snapshots so that the changes happening as part of the passed transaction can be correctly
+     * committed or rolled back.
+     * This function should be called every time the journal is about to change its internal state as part of a transaction.
+     * However, only the first snapshot taken of that depth will be taken.
+     */
+    public void updateSnapshots(TransactionContext transaction) {
+        int currentDepth = transaction.depth();
+
+        snapshots.ensureCapacity(currentDepth);
+        for (int i = snapshots.size(); i <= currentDepth; i++) {
+            snapshots.add(null);
+        }
+
+        if (snapshots.get(currentDepth) == null) {
+            T snapshot = createSnapshot();
+            Objects.requireNonNull(snapshot, "Snapshot may not be null!");
+            snapshots.set(currentDepth, snapshot);
+
+            // This is a special case where we need to cast to access the add journalToCommit method.
+            // This should not be used as a usage example for your implementations
+            // with exception to the updateSnapshots method though you should just be able to call super.updateSnapshots to avoid
+            // doing it yourself when necessary.
+            if (transaction instanceof Transaction resolvedTransaction)
+                resolvedTransaction.addClosingJournal(this);
```

Remove the useless comment, and write a cast instead of an instanceof:

```suggestion
            ((Transaction) transaction).addClosingJournal(this);
```

---

We should inform someone looking at this as to WHY we do this and why they should not, as we tell them not to, then blatantly do so since it is necessary.

 As for the casting, it makes no difference really, I'll change it, but I am more curious as to why?

---

Your updated comment is also fine. Essentially I'd rather crash if someone ever sends a custom TransactionContext implementation, rather than silently do nothing.

---

We could (and likely should) seal the `TransactionContext` permitting `Transaction`

---

Yeah we should.